File: day04.c
#include <stdio.h>

int main(void){
    int part1=0, part2=0, N=0;
    char T[150][150];
    FILE *f = fopen("data/input04.txt","r");
    while(fgets(T[N],150,f)){ N++; }
  
    int C(int i,int j,char c){return i>=0 && i<N && j>=0 && j<N && T[i][j]==c;}
    int XMAS(int i, int j, int di, int dj){
        return C(i,j,'X') && C(i+di,j+dj,'M') && C(i+2*di,j+2*dj,'A') && C(i+3*di,j+3*dj,'S');
    }
    int MAS(int i, int j, int di, int dj){
        int r1 = C(i-di,j-dj,'M') && C(i,j,'A') && C(i+di,j+dj,'S');
        int r2 = C(i+di,j+dj,'M') && C(i,j,'A') && C(i-di,j-dj,'S');
        return r1||r2;
    }

    for(int i=0; i<N; i++){
        for(int j=0; j<N; j++){
            for(int di=-1; di<=1; di++){
                for(int dj=-1; dj<=1; dj++){
                    part1+=XMAS(i,j,di,dj);
                }
            }
            part2+=MAS(i,j,1,1) && MAS(i,j,-1,1);
        }
    }
    printf("%d %d\n",part1,part2);
    return 0;
}

File: day04.py
part1,part2=0,0
with open("data/input04.txt") as f:
    L=[list(x) for x in f.read().strip().split("\n")]
r,c=len(L),len(L[0])
H=("".join(L[i][j+k] for k in range(4)) for i in range(r) for j in range(c-3))
V=("".join(L[i+k][j] for k in range(4)) for i in range(r-3) for j in range(c))
D1=("".join(L[i+k][j+k] for k in range(4)) for i in range(r-3) for j in range(c-3))
D2=("".join(L[i+k][j+3-k] for k in range(4)) for i in range(r-3) for j in range(c-3))
part1+=sum(w=="XMAS" or w=="SAMX" for w in H)
part1+=sum(w=="XMAS" or w=="SAMX" for w in V)
part1+=sum(w=="XMAS" or w=="SAMX" for w in D1)
part1+=sum(w=="XMAS" or w=="SAMX" for w in D2)
for i in range(r-2):
    for j in range(c-2):
        w1="".join([L[i+k][j+k] for k in range(3)])
        w2="".join([L[i+k][j+2-k] for k in range(3)])
        if (w1=="MAS"or w1=="SAM") and (w2=="MAS" or w2=="SAM"): part2+=1
print(part1,part2)